crates/types/src/string.rs

```diff
@@ -201,16 +180,154 @@ impl StringBuilder {
+            return;
+        }
+        let needed_space =
+            if self.inner.data.is_null() { cap + 1 } else { cap - remaining };
+        let new_capacity = (self.cap + needed_space)
+            .checked_next_power_of_two()
+            // in the rare case that the next power of two overflows we should allocate the minimal
+            // amount to avoid OOM problems
+            .unwrap_or(self.cap + needed_space);
+        unsafe { self.reserve_raw(new_capacity) };
+        self.cap = new_capacity;
+    }
+
+    /// Reserve space for exactly N more bytes.
+    ///
+    /// Does not allocate if enough space is available.
+    pub fn reserve_exact(&mut self, cap: usize) {
+        if cap == 0 {
+            return;
+        }
+        let remaining = self.remaining_capacity();
+        if remaining > cap {
+            return;
+        }
+        let needed_space =
+            if self.inner.data.is_null() { cap + 1 } else { cap - remaining };
+        let new_capacity = self.cap + needed_space;
+        unsafe { self.reserve_raw(new_capacity) };
+        self.cap = new_capacity;
+    }
+
+    /// Reserve at least "by" bytes.
+    ///
+    /// This is marked unsafe as the allocation size should never be zero for portability reasons.
+    /// A check isn't included as all call sites should already handle the case.
+    unsafe fn reserve_raw(&mut self, by: usize) {
+        let ptr =
+            unsafe { libc::realloc(self.inner.data as *mut ffi::c_void, by) };
+        // realloc may return null if it is unable to allocate the requested memory
+        if ptr.is_null() {
+            unable_to_alloc_memory();
+        }
+        self.inner.data = ptr as *mut ffi::c_char;
+    }
+
     /// Build the `String`.
     #[inline]
-    pub fn finish(self) -> String {
-        let s = String { data: self.inner.data, len: self.inner.len };
+    pub fn finish(mut self) -> String {
+        let s = &mut self.inner;
+        // TODO: since strings constructed via nvim-oxi are leaked, we should at least leak the
+        // minimal amount possible. Once a more sophisticated solution is found this should be
+        // removed since this can cause an allocation if the data needs to be moved to a new
+        // address.
+        if !s.data.is_null() {
+            if s.is_empty() {
+                unsafe { libc::free(s.data as *mut ffi::c_void) };
+                // still call `StringBuilder::finish_unchecked` for the debug assertions and
+                // mem::forget
+                _ = self.finish_unchecked();
+                return String::new();
+            } else if s.len() + 1 < self.cap {
+                let ptr = unsafe {
+                    libc::realloc(s.data as *mut ffi::c_void, s.len() + 1)
+                };
+                if ptr.is_null() {
+                    unable_to_alloc_memory();
+                }
+                s.data = ptr as *mut ffi::c_char;
+            }
+        };
+
+        self.finish_unchecked()
+    }
```

Comment:
Strings being leaked is a separate problem which should hopefully be resolved Soon™. I think we should just have a single `finish()` function.